File: ios/Sources/App/Services/CommandHistoryStore.swift
import Foundation

struct CommandHistoryRecord: Codable, Identifiable, Equatable {
    let id: UUID
    let hostID: UUID
    let command: String
    let timestamp: Date
    let status: String? // success, error, timeout
    var isFavorite: Bool

    var displayCommand: String {
        let trimmed = command.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.count > 80 ? String(trimmed.prefix(77)) + "…" : trimmed
    }
}

@MainActor
final class CommandHistoryStore: ObservableObject {
    static let shared = CommandHistoryStore()
    @Published private(set) var records: [CommandHistoryRecord] = []

    private let fileURL: URL
    private let maxRecords = 1000

    private init() {
        let directory = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("TermVault", isDirectory: true)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        fileURL = directory.appendingPathComponent("command-history.json")
        if let data = try? Data(contentsOf: fileURL),
           let decoded = try? JSONDecoder().decode([CommandHistoryRecord].self, from: data) {
            records = decoded
        }
    }

    func record(_ command: String, hostID: UUID, status: String? = nil) {
        guard !command.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }
        records.insert(CommandHistoryRecord(
            id: UUID(),
            hostID: hostID,
            command: command,
            timestamp: Date(),
            status: status,
            isFavorite: false
        ), at: 0)
        records = Array(records.prefix(maxRecords))
        save()
    }

    func toggleFavorite(_ record: CommandHistoryRecord) {
        guard let index = records.firstIndex(where: { $0.id == record.id }) else { return }
        records[index].isFavorite.toggle()
        save()
    }

    func delete(_ record: CommandHistoryRecord) {
        records.removeAll { $0.id == record.id }
        save()
    }

    func search(_ query: String, hostID: UUID? = nil) -> [CommandHistoryRecord] {
        var results = records.filter { $0.command.localizedCaseInsensitiveContains(query) }
        if let hostID = hostID {
            results = results.filter { $0.hostID == hostID }
        }
        return results
    }

    func favorites(for hostID: UUID? = nil) -> [CommandHistoryRecord] {
        var results = records.filter { $0.isFavorite }
        if let hostID = hostID {
            results = results.filter { $0.hostID == hostID }
        }
        return results
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(records) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }
}

File: android/app/src/main/kotlin/dev/termvault/app/history/CommandHistoryStore.kt
package dev.termvault.app.history

import dev.termvault.app.data.db.CommandHistoryDao
import dev.termvault.app.data.db.CommandHistoryEntity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.stateIn
import java.util.UUID

private const val MAX_RECORDS = 1000

/** Mirrors iOS `CommandHistoryStore`, backed by Room instead of a JSON file. */
class CommandHistoryStore(private val dao: CommandHistoryDao, scope: CoroutineScope) {
    val records: StateFlow<List<CommandHistoryEntity>> =
        dao.observeAll().stateIn(scope, SharingStarted.Eagerly, emptyList())

    suspend fun record(command: String, hostId: UUID, status: String? = null) {
        if (command.isBlank()) return
        dao.insert(CommandHistoryEntity(hostId = hostId, command = command, status = status))
        dao.trimTo(MAX_RECORDS)
    }

    suspend fun toggleFavorite(record: CommandHistoryEntity) {
        dao.upsert(record.copy(isFavorite = !record.isFavorite))
    }

    suspend fun delete(record: CommandHistoryEntity) {
        dao.delete(record)
    }

    fun search(query: String, hostId: UUID? = null): List<CommandHistoryEntity> =
        records.value.filter { it.command.contains(query, ignoreCase = true) && (hostId == null || it.hostId == hostId) }

    fun favorites(hostId: UUID? = null): List<CommandHistoryEntity> =
        records.value.filter { it.isFavorite && (hostId == null || it.hostId == hostId) }
}
